[AG-11882] Fix(master-detail): allow master rows to update whether they display an expansion button

sdwvit opened the pull request:
Enable dynamic expansion button visibility for master rows via isRowMaster callback

Local example SSRM: https://plnkr.co/edit/zAXiCJ2NIM0VP8mM?open=main.ts&preview
Local example CSRM: https://plnkr.co/edit/tXPXIpw4Txhnovwe?open=main.ts&preview

Fixes [AG-11882](https://ag-grid.atlassian.net/browse/AG-11882)
Fixes [AG-16107](https://ag-grid.atlassian.net/browse/AG-16107)

sdwvit committed:
refactor row rendering logic for master-detail functionality

 - renamed createRowCon to createRowCtrl for clarity
 - updated row control creation to respect master-detail state
 - added logic to refresh rows based on isRowMaster callback
 - introduced shouldSetMasterDetailFlag method for better readability

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

seanlandsman committed:
DASH-11 Auto grid bump and publish trigger

sdwvit committed:
Merge branch 'latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

sdwvit committed:
Refactor master-detail logic for dynamic row updates

 - Remove redundant shouldSetMasterDetailFlag method.
 - Simplify master-detail flag checks in updateDataIntoRowNode.
 - Enhance refreshRow to ensure correct rendering based on row master status.
 - Update setDataCommon to allow dynamic master flag during runtime.

sdwvit committed:
- Update comment for clarity

SalvatorePreviti commented on packages/ag-grid-enterprise/src/serverSideRowModel/blocks/blockUtils.ts:
```diff
@@ -197,12 +198,14 @@ export class BlockUtils extends BeanStub implements NamedBean {
-        if ((treeData || !group) && this.gos.get('masterDetail')) {
+        const isTreeOrNotAGroup = treeData || !group;
+        const isMasterDetail = gos.get('masterDetail');
+        if (isTreeOrNotAGroup && isMasterDetail) {
```
there is a better way - group nodes or tree data filler nodes cannot be masters, but tree data groups with data can be master.
Tree data can have also filler nodes, nodes without data.
I think the right way is to check if (rowNode.data) then it can be master, else it can't be

SalvatorePreviti commented on packages/ag-grid-community/src/rendering/row/rowCtrl.ts:
```diff
@@ -880,18 +880,35 @@ export class RowCtrl extends BeanStub<RowCtrlEvent> {
 
     public refreshRow(params: RefreshRowsParams & { newData?: boolean }): void {
-        // if the row is rendered incorrectly, as the requirements for whether this is a FW row have changed, we force re-render this row.
+        const gos = this.gos;
```
is really this the right place to do this? Besides the fact that seems very expensive, is not there a better place? this is also calling redraw if master field didn't change - i think this logic is already in place somewhere to refresh the row if master flag changes?

SalvatorePreviti commented on packages/ag-grid-community/src/entities/rowNode.ts:
```diff
@@ -292,21 +292,37 @@ export class RowNode<TData = any>
-
         this.data = data;
+
+        const hasMasterFunc = gos.get('isRowMaster');
```
I know this function is exposed to the end user, this adds a level of complexity for sure ... but we are invoking this callback twice in this way, in MasterDetailService and here

- or remove it from there if this is the case and completely change that implementation 
- or do it instead in SSRM when data changes

SalvatorePreviti commented on packages/ag-grid-community/src/rendering/row/rowCtrl.ts:
```diff
@@ -880,18 +880,35 @@ export class RowCtrl extends BeanStub<RowCtrlEvent> {
+        const gos = this.gos;
+        const hasRowMasterFunc = gos.get('isRowMaster');
+        const isRowMaster = gos.get('masterDetail');
+        const redraw = () => this.beans.rowRenderer.redrawRow(this.rowNode);
```
this creates a new object, no need to create a new function here

SalvatorePreviti commented on packages/ag-grid-community/src/rendering/row/rowCtrl.ts:
```diff
@@ -880,18 +880,35 @@ export class RowCtrl extends BeanStub<RowCtrlEvent> {
+         * First part is in rowNode.ts where row.master is set on same conditions.
+         */
+        if (hasRowMasterFunc && isRowMaster) {
+            this.getAllCellCtrls().forEach((cellCtrl) => cellCtrl.refreshCell(params));
```
sonarQube new rules says: no .forEach - we are removing all .forEach as we find them - let's not add new .forEach - better to use for loops

SalvatorePreviti commented:
This would really benefit from behavioural unit tests both for CSRM and SSRM also to verify we don't invoke isMaster callback multiple times.
This is much more impactful than it seems

Currently tests are failing

sdwvit committed:
Merge branch 'latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

sdwvit committed:
Refactor master-detail row update logic

 - Replace updateData with updateDataInternal in multiple files for consistency
 - Simplify master determination logic in setMaster method
 - Remove unnecessary comments and improve clarity in rowCtrl and rowNode
 - Ensure internal updates do not trigger master checks unnecessarily

sdwvit committed:
Update rowNode data update method for refresh compatibility

 - Changed from internal to external data update to trigger refreshes

SalvatorePreviti commented on packages/ag-grid-community/src/entities/rowNode.ts:
```diff
@@ -291,22 +291,30 @@ export class RowNode<TData = any>
         this.data = data;
+
+        if (!internal) {
+            this.beans.masterDetailSvc?.setMaster(this, false, true);
```
Just an idea for clarity:

- instead of having an internal flag, I would just repeat `this.beans.masterDetailSvc?.setMaster(this, false, true);` in the two public facing methods
- instead of creating a new `updateDataInternal` method that is not clear what it does, we can just make `setDataCommon` public maybe?
- we need to add documentation that internally updateData and setData should NOT be called and instead we should call setDataCommon

However - we could consider to change approach altogether and use a flag in the rowModel - see other comments

SalvatorePreviti replied:
we have also other calls to updateData internally that can create problems and we need to check
- the clientSideChildrenTreeNodeManager uses updateData and this needs to call the internal version instead
- SSRM is for sure already setting this flag and calling the dispatch event - if we call updateData as it is now aren't we going to have two calls to isMaster callback?
- we need to check viewPortRowModel and staticPinnedRowModel too - other places where updateData is called

SalvatorePreviti commented on packages/ag-grid-community/src/entities/rowNode.ts:
```diff
@@ -291,22 +291,30 @@ export class RowNode<TData = any>
     }
 
-    private setDataCommon(data: TData, update: boolean): void {
+    public updateDataInternal(data: TData): void {
```
This should work now - in theory besides checking all the places we call updateData - I don't know if it would be better and cleaner to instead add a `readonly updatingData?: boolean` in the IRowModel interface and check in the masterDetailService.setMaster if this flag is on, if so skip setting the master? Delegating the responsibility away from the RowNode class and from who calls those methods

sdwvit committed:
Refactor row data update methods for consistency

 - Replace updateDataInternal with updateData in multiple files for uniformity.
 - Update master-detail service to listen for rowNodeDataChanged events.
 - Simplify setDataCommon method by removing internal flag handling.

sdwvit committed:
Allow conditional refresh of master-detail model

 - Added skipMasterDetailsModelRefresh parameter to control master detail refresh.
 - Updated row data update logic to respect the new parameter.

sdwvit committed:
- Always call masterDetailService.refreshModel() regardless of params.skipMasterDetailsModelRefresh.
 - Remove skipMasterDetailsModelRefresh property from RefreshModelParams interface.
 - Enhance event listener management for master-detail updates.

sdwvit committed:
Refactor event listener management in masterDetailService

sdwvit committed:
if for some reason destructors is prepopulated

sdwvit committed:
apply suggestion from PR review

sdwvit committed:
apply suggestion from PR review

sdwvit committed:
Merge branch 'latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

sdwvit committed:
yarn

sdwvit committed:
revert some changes

sdwvit committed:
Merge branch 'latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

# Conflicts:
#	yarn.lock

sdwvit committed:
revert some changes

SalvatorePreviti commented on packages/ag-grid-enterprise/src/masterDetail/masterDetailService.ts:
```diff
@@ -50,59 +76,58 @@ export class MasterDetailService extends BeanStub implements NamedBean, IMasterD
 
         const gos = this.gos;
-        const isRowMaster = gos.get('isRowMaster');
+        const isRowMaster = gos.get('isRowMaster') as ((data: any) => unknown) | undefined;
```
minor: is this cast needed?

sdwvit replied:
good spot, not needed

sdwvit committed:
pr comment

sdwvit committed:
try mitigate mem issues

sdwvit commented on packages/ag-grid-enterprise/src/masterDetail/masterDetailService.ts:
```diff
@@ -50,59 +77,64 @@ export class MasterDetailService extends BeanStub implements NamedBean, IMasterD
+                delete this.store[row.detailNode.id!];
+                (beans.ssrmBlockUtils as BlockUtils)?.destroyRowNode(row.detailNode);
+                row.detailNode = undefined;
             }
```
@SalvatorePreviti

Commit:
lint

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Merge tag 'latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

Latest commit to pass GitHub Actions workflow on latest branch.

Signed-off-by: Victor Musienko <victor.m@noibu.com>

# Conflicts:
#	.github/workflows/beta-publish.yml

Commit:
Merge branch 'refs/heads/latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

# Conflicts:
#	packages/ag-grid-enterprise/src/masterDetail/masterDetailService.ts

Commit:
revert stuff

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
- Add event listener for masterChanged to trigger destruction
 - Remove unnecessary detail node deletion logic

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
upd

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
add destructor

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
add a comment

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Merge branch 'refs/heads/latest' into sdwvit/fix/master-detail/allow-master-rows-to-update-whether-they-display-an-expansion-button/AG-11882

Commit:
remove unused import

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
refactor per pr comment

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
refactor per pr comment

Signed-off-by: Victor Musienko <victor.m@noibu.com>
